### PR TITLE
Fix config bugs

### PR DESCRIPTION
**Pull request by anaesqueda**
Fixes Issue #17

> It's a good idea to open an issue first for discussion.

- [ ] All Tests pass

### CODE REVIEW

File: utils/utils.py
Comment (pkenjora) [inline utils/utils.py:1]: Recommend moving this method to Configuration class. Avoids new folder plus file and logically its a member of configuration since it depends on that class.

File: main.py [hunk 1]
```diff
@@ -40,156 +38,130 @@
 from helpers.vertex_ai_service import LLMParameters
 from helpers.bq_service import BigQueryService
 from configuration import Configuration
+from utils.utils import parse_args, build_abcd_params_config
```
Comment (pkenjora) [inline main.py:41]: Move these methods to Configuration class, remove extra import.

File: main.py [hunk 2]
```diff
@@ -40,156 +38,130 @@
...
-
-    return brand_assessment
+  Args:
+    arg_list: A list of command line arguments
+
+  """
+
+  args = parse_args(arg_list)
+
+  config = build_abcd_params_config(args)
```
Comment (pkenjora) [inline main.py:151]: Change to config.from_args() and move both: parse_args(arg_list) and build_abcd_params_config(args) into that class.

File: [GitHub]_ABCDs_Detector.ipynb [hunk 3]
```diff
@@ -182,47 +171,49 @@
...
+        "id": "3KlVPryttrZr"
+      },
+      "outputs": [],
       "source": [
         "!git clone -b paul_changes https://github.com/google-marketing-solutions/abcds-detector\n",
         "!python3 -m pip install -r \"abcds-detector/requirements.txt\"\n",
         "%cd /content/abcds-detector\n",
         "\n",
         "from configuration import Configuration\n",
         "config = Configuration()"
```
Comment (pkenjora) [inline [GitHub]_ABCDs_Detector.ipynb:198]: Move to own section to make the configuration more explicit.  Users running Install may not realize Configuration is also being set. The following code into the notebook will work:
```
    {   
      "cell_type": "markdown",
      "source": [
        "# Install ABCD Detector Code"
      ],
      "metadata": {
        "id": "m-VVH3V36y8y"
      } 
    },  
    {   
      "cell_type": "code",
      "source": [ 
        "!git clone https://github.com/google-marketing-solutions/abcds-detector\n",
        "!python3 -m pip install -r \"abcds-detector/requirements.txt\""
      ],
      "metadata": {
        "id": "3KlVPryttrZr"
      },
      "execution_count": null,
      "outputs": []
    },  
    {   
      "cell_type": "markdown",
      "source": [
        "# Initialize ABCDs Detector Configuration"
      ],
      "metadata": {
        "id": "m-VVH3W36y8y"
      }
    },
    {   
      "cell_type": "code",
      "source": [ 
        "%cd /content/abcds-detector\n",
        "from configuration import Configuration\n",
        "config = Configuration()"
      ],
      "metadata": {
        "id": "3KlVPrWttrZr"
      },
      "execution_count": null,
      "outputs": []
    },
```

File: tests/test_abcs_parameters.py [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+  args = ArgsMock(project_id="tighlock-test", project_zone="us-central1",
+    bucket_name="abcd-detector-input", knowledge_graph_api_key="asdfds", bigquery_dataset="abcd_detector_ds", bigquery_table="my_table",
+    assessment_file="", use_annotations=True, use_llms=True, verbose=True, annotation_path="",video_uris="gs://abcd-detector-input/Google/videos/",
+    brand_name="Google", brand_variations="Google,google", branded_products="Google pixel, Google pixel buds, Google pixel watch",
+    branded_products_categories="phone, watch, buds", branded_call_to_actions="buy it!, buy", early_time_seconds=5, confidence_threshold=0.5,
+    face_surface_threshold=0.5, logo_size_threshold=0.5, avg_shot_duration_seconds=3, dynamic_cutoff_ms=3000, llm_name="asd",
+    video_size_limit_mb=50, max_output_tokens=8000, temperature=1, top_k=0.1, top_p=0.1
+  )
+
+  config = build_abcd_params_config(args)
```
Comment (pkenjora) [inline tests/test_abcs_parameters.py:144]: Move to method of class.

File: tests/test_abcs_parameters.py [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+    # set model
+    self.llm_name = llm_name
+    self.video_size_limit_mb = video_size_limit_mb
+    self.max_output_tokens = max_output_tokens
+    self.temperature = temperature
+    self.top_p = top_p
+    self.top_k = top_k
+
+
+def test_not_empty_abcd_params():
```
Comment (pkenjora) [inline tests/test_abcs_parameters.py:132]: Add test for setting ONLY the project, bucket, and KEY.
The configuration by default should evaluate a Google branded iphone video in a public samples folder. This is required for easy demos.

File: [GitHub]_ABCDs_Detector.ipynb [hunk 6]
```diff
@@ -182,47 +171,49 @@
...
     },
     {
       "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "3KlVPryttrZr"
+      },
+      "outputs": [],
       "source": [
         "!git clone -b paul_changes https://github.com/google-marketing-solutions/abcds-detector\n",
```
Comment (pkenjora) [inline [GitHub]_ABCDs_Detector.ipynb:193]: Remove -b paul_changes, this was an artifact left over from testing the release, it was not an issue while the branch and head were in sync, however now that additional changes are in, this should be removed.

File: main.py [hunk 7]
```diff
@@ -128,6 +125,7 @@ def execute_abcd_assessment_for_videos(config: Configuration):
     brand_assessment.get("video_assessments").append(video_assessment)
 
     if config.bq_table_name:
+      bq_service = BigQueryService(config.project_id)
```
Comment (pkenjora) [inline main.py:128]: Move this into store_in_bq, leverage config to pass parameter.
Exposing the internals of store_in_bq adds BigQueryService dependency here. Anyone modifying the code for AWS for example will have to remove 2 dependencies instead of 1.

**Review by pkenjora — CHANGES_REQUESTED** [review]
Please move dependency lower down.

File: tests/test_abcs_parameters.py (hunk 4 above)
Comment (pkenjora) [reply tests/test_abcs_parameters.py:144]: For example, mock sys.argv

```
import argparse
from unittest import TestCase, mock

def main():
    parser = argparse.ArgumentParser()
    parser.add_argument('--name', required=True)
    args = parser.parse_args()
    print(f'Hello {[args.name](http://args.name/)}')

class TestArgparse(TestCase):
    def test_main(self):
        with mock.patch('sys.argv', ['my_script.py', '--name', 'John']):
            config = Configuration()
           config.from_args()
          # assert values
```

File: tests/test_abcs_parameters.py [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+#    limitations under the License.
+#
+###########################################################################
+
+""" Module to test ABCD parameters """
+
+from utils.utils import build_abcd_params_config
+
+
+class ArgsMock():
```
Comment (smcjones) [inline tests/test_abcs_parameters.py:26]: consider [dataclass](https://docs.python.org/3/library/dataclasses.html) here. It will remove the need to add an __init__ method.

File: tests/test_abcs_parameters.py (hunk 4 above)
Comment (smcjones) [reply tests/test_abcs_parameters.py:144]: Would recommend keeping as-is to follow [best practices](https://engdoc.corp.google.com/eng/doc/devguide/py/style/index.md?cl=head#:~:text=Never%20use%20staticmethod%20unless%20forced%20to%20in%20order%20to%20integrate%20with%20an%20API%20defined%20in%20an%20existing%20library.%20Write%20a%20module%2Dlevel%20function%20instead)

File: tests/test_abcs_parameters.py [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+    self.max_output_tokens = max_output_tokens
+    self.temperature = temperature
+    self.top_p = top_p
+    self.top_k = top_k
+
+
+def test_not_empty_abcd_params():
+  """Tests that all brand parameters are provided"""
+
+  args = ArgsMock(project_id="tighlock-test", project_zone="us-central1",
```
Comment (smcjones) [inline tests/test_abcs_parameters.py:135]: Suggestion to make this a [fixture](https://docs.pytest.org/en/6.2.x/fixture.html) - optional but can help to clean up some of the setup steps from the test. Not necessary until you have more tests.

File: main.py (hunk 1 above)
Comment (smcjones) [inline main.py:41]: `utils.utils` sounds a bit repetitive. Can this be moved up one level so it can be imported as `utils`?
```suggestion
from utils import parse_args, build_abcd_params_config
```

File: utils/utils.py [hunk 10]
```diff
@@ -0,0 +1,161 @@
...
+        brand_name=args.brand_name,
+        brand_variations=args.brand_variations,
+        products=args.branded_products,
+        products_categories=args.branded_products_categories,
+        call_to_actions=args.branded_call_to_actions,
+    )
+    return config
+
+
+def parse_args(arg_list: list[str] | None = None) -> None:
```
Comment (smcjones) [inline utils/utils.py:61]: No action for this one, but I'm partial to [click](https://click.palletsprojects.com/en/stable/) over argparse for testability and usability.

File: main.py (hunk 2 above)
Comment (smcjones) [reply main.py:151]: it is in-line with best practices as-is so would not recommend doing this.

File: main.py (hunk 7 above)
Comment (smcjones) [reply main.py:128]: `store_in_bq` would still be a dependency. Moving this into `store_in_bq` just shifts the dependency down one level. I recommend leaving this as-is until the need to refactor comes up. [Premature optimization is the root of all evil](https://m.xkcd.com/1691/).

File: [GitHub]_ABCDs_Detector.ipynb (hunk 6 above)
Comment (smcjones) [inline [GitHub]_ABCDs_Detector.ipynb:193]:
```suggestion
        "!git clone -b main https://github.com/google-marketing-solutions/abcds-detector\n",
```

File: tests/test_abcs_parameters.py (hunk 4 above)
Comment (anaesqueda) [reply tests/test_abcs_parameters.py:144]: +1 we need those functions separate from the Config class to be able to test how this object is being built.

File: tests/test_abcs_parameters.py (hunk 5 above)
Comment (anaesqueda) [reply tests/test_abcs_parameters.py:132]: +1, this test suite will contain happy path and edge cases. We still need to test brand details params, as Google brand details and not set by default in this class. It might be for the videos for not brand name, products, etc.

File: main.py (hunk 7 above)
Comment (anaesqueda) [reply main.py:128]: +1 this is still a dependency. If this gets removed, or replaced, it would be the whole condition.

File: tests/test_abcs_parameters.py (hunk 8 above)
Comment (anaesqueda) [reply tests/test_abcs_parameters.py:26]: +1!

File: tests/test_abcs_parameters.py (hunk 9 above)
Comment (anaesqueda) [reply tests/test_abcs_parameters.py:135]: +1 Doing that as part of the comprehensive testing suite on future CLs

File: main.py (hunk 1 above)
Comment (anaesqueda) [reply main.py:41]: Makes sense, removing the folder.

File: tests/test_abcs_parameters.py (hunk 4 above)
Comment (pkenjora) [reply tests/test_abcs_parameters.py:144]: That may be the wrong best practice, there is no static method being suggested here.
On the contrary it would have access to self to set the parameters.

File: [GitHub]_ABCDs_Detector.ipynb (hunk 3 above)
Comment (anaesqueda) [reply [GitHub]_ABCDs_Detector.ipynb:198]: The colab didn't have any changes other than a method name change. I would not make more changes to this unless something is actually broken.

File: tests/test_abcs_parameters.py (hunk 4 above)
Comment (smcjones) [reply tests/test_abcs_parameters.py:144]: This is the best practices I was referring to, so maybe we can consolidate to this conversation.

This is written as a builder function that creates the configuration class and populates it in one go. For it to have the same function inside of a class, it would have to be a `@classmethod`.

File: [GitHub]_ABCDs_Detector.ipynb [hunk 11]
```diff
@@ -278,7 +269,7 @@
         "BRANDED_PRODUCT_CATEGORIES = \"phone, watch, buds\"  # @param {type:\"string\"}\n",
         "BRANDED_CALL_TO_ACTIONS = \"buy it!\"  # @param {type:\"string\"}\n",
         "\n",
-        "config.set_brand(\n",
+        "config.set_brand_details(\n",
         "    name = BRAND_NAME,\n",
```
Comment (smcjones) [inline [GitHub]_ABCDs_Detector.ipynb:273]:
```suggestion
        "    brand_name = BRAND_NAME,\n",
```

File: [GitHub]_ABCDs_Detector.ipynb [hunk 12]
```diff
@@ -278,7 +269,7 @@
...
         "    variations = BRAND_VARIATIONS,\n",
```
Comment (smcjones) [inline [GitHub]_ABCDs_Detector.ipynb:274]:
```suggestion
        "    brand_variations = BRAND_VARIATIONS,\n",
```

**Review by smcjones — COMMENTED** [review]
I ran the colab and needed to make these changes to get it to run

File: main.py (hunk 2 above)
Comment (smcjones) [reply main.py:151]: Responded in the other thread.

**Review by smcjones — APPROVED** [review]
LGTM

File: tests/test_abcs_parameters.py (hunk 4 above)
Comment (pkenjora) [reply tests/test_abcs_parameters.py]: Nothing prevents it from being an instance method, as a technical call.  Ana's call as an implementation call.

**Review by pkenjora — APPROVED** [review]

File: tests/test_abcs_parameters.py (hunk 4 above)
Comment (smcjones) [reply tests/test_abcs_parameters.py]: I prefer the functional approach because it's self documenting, while a class that needs to have a method called after instantiation is less obvious to me and feels very error-prone. I'd prefer a `@classmethod` to a regular method. I'd probably recommend a regular method get refactored to a function or `@classmethod`, but agree that it's the code owner's choice so long as there is a consistency in style.

**Review by jyotheeshwar — APPROVED** [review]
